### PR TITLE
[Doc] Refactored contribution guidelines

### PR DESCRIPTION
**Pull request by Vissidarte-Herman**
One section shy of completion: which is the "file a good pr" section.

Related issue = #

### CODE REVIEW

File: docs/lang/articles/contribution/contributor_guide.md [hunk 1]
```diff
@@ -4,296 +4,324 @@ sidebar_position: 1
...
+- Documentation source files are hosted under [docs/](https://github.com/taichi-dev/taichi/blob/master/docs/).
+- We use GitHub Flavored Markdown (GFM) and [Docusaurus](https://docusaurus.io/) to build our documentation site. For information on the supported Markdown syntax, see the  [Documentation Writing Guide](./doc_writing). 
+- When it comes to writing, we adhere to the [Google Developer Documentation Style Guide](https://developers.google.com/style/).
+- For instruction on setting up a local server and previewing your updated documentation in real-time, see the [Local Development](https://github.com/taichi-dev/docs.taichi.graphics#local-development).
+
+## Add test cases for testing your local changes
+
+We strongly recommend that you write your own test cases to cover corner cases for your codes before filing a PR.
+
+- To write a Python test case, see the [Workflow for writing a Python test](/write_test).
```
Comment (ailzhang) [inline docs/lang/articles/contribution/contributor_guide.md:107]: possibly missing a `.` here? (which caused the deployment failure IIUC :P

File: docs/lang/articles/contribution/contributor_guide.md [hunk 2]
```diff
@@ -4,296 +4,324 @@ sidebar_position: 1
...
 
 Frequently used tags:
 
-- `[CPU]`, `[CUDA]`, `[Metal]`, `[Vulkan]`, `[OpenGL]`: backends;
-- `[LLVM]`: the LLVM backend shared by CPUs and CUDA;
-- `[Lang]`: frontend language features, including syntax sugars;
-- `[Std]`: standard library, e.g., `ti.Matrix` and `ti.Vector`;
-- `[Sparse]`: sparse computation;
-- `[IR]`: intermediate representation;
+- `[CPU]`, `[CUDA]`, `[Metal]`, `[Vulkan]`, `[OpenGL]`: Backends;
```
Comment (k-ye) [inline docs/lang/articles/contribution/contributor_guide.md:227]: Let's just keep "CI", "CUDA", "IR", "Lang", "Refactor".

File: docs/lang/articles/contribution/contributor_guide.md [hunk 3]
```diff
@@ -4,296 +4,324 @@ sidebar_position: 1
...
-  should use tags with **everything in lowercase letters**, e.g.,
-  `[metal]`, `[vulkan]`, `[ir]`, `[lang]`, `[cli]`.
-- Because of the way the release changelog is generated, there
-  should be **at most one capitalized tag** in a PR title to prevent
-  duplicate PR highlights. For example,
-  `[GUI] [Mac] Support modifier keys` ([#1189](https://github.com/taichi-dev/taichi/pull/1189))
-  is an improper tag choice, and we
-  should have used `[gui] [Mac] Support modifier keys in GUI` instead.
-  Please capitalize the tag that is the *most* relevant to the PR.
+We do appreciate all kinds of contributions, yet we do not intend to expose the details of every PR to the end-users. Therefore the changelog helps distinguish *what the user should know* from *what the*
```
Comment (k-ye) [inline docs/lang/articles/contribution/contributor_guide.md:269]: Hmm, is this still valid... I.e. do we still distinguish upper-case and lower-case PR tags in our release notes? @strongoier @ailzhang

File: docs/lang/articles/contribution/contributor_guide.md [hunk 4]
```diff
@@ -4,296 +4,324 @@ sidebar_position: 1
...
+- Other PRs (underlying development or intermediate implementation) should use tags **in lowercase letters**, for example`[metal]`, `[vulkan]`, `[ir]`, `[lang]`.
+
+- Because of the way the release changelog is generated, there should be **at most one capitalized tag** in a PR title to prevent duplicate PR highlights. For example, `[GUI] [Mac] Support modifier keys` ([#1189](https://github.com/taichi-dev/taichi/pull/1189))is an improper tag choice, and we should have used `[gui] [Mac] Support modifier keys in GUI` instead.
+
+Please capitalize the most relevant tag to your PR.
+
 :::
 
-## Testing
+### **Use continuous integration**
```
Comment (k-ye) [inline docs/lang/articles/contribution/contributor_guide.md:283]: CI is set up by ourselves and runs automatically on each commit. In most cases contributors wouldn't need to worry about it.

```suggestion
### Continuous Integration (CI)
```

File: docs/lang/articles/contribution/contributor_guide.md [hunk 5]
```diff
@@ -4,296 +4,324 @@ sidebar_position: 1
...
 
-Tests should be added to [tests/](https://github.com/taichi-dev/taichi/blob/master/tests/). We
-have both Python tests and C++ tests.
+- Continuous Integration (CI) will **build** and **test** your commits in a specific PR in multiple environments.
 
-### Python tests
+- CI is triggered each time you push a commit to an open PR.
+
+- A tick on the left-hand side of a commit hash means that CI passes, whilst a cross means that CI fails.
```
Comment (k-ye) [inline docs/lang/articles/contribution/contributor_guide.md:289]: Hmm, consider ✔️ and ❌?

File: docs/lang/articles/contribution/contributor_guide.md [hunk 6]
```diff
@@ -4,296 +4,324 @@ sidebar_position: 1
...
+
+### **PR reviewing & merging**
+
+- Please try to follow these tips from Google:
+  - [Code Health: Understanding Code In](https://testing.googleblog.com/2018/05/code-health-understanding-code-in-review.html)
```
Comment (k-ye) [inline docs/lang/articles/contribution/contributor_guide.md:294]: Typo: the original content only had two list items. It should have been

* Code Health: Understanding Code In Review
* Code Health: Respectful Reviews == Useful Reviews

https://github.com/taichi-dev/taichi/blob/9f0d9f505cfdb9aec6317fb394ef88d01e9cb2d2/docs/lang/articles/contribution/contributor_guide.md?plain=1#L141-L147

File: docs/lang/articles/contribution/contributor_guide.md [hunk 7]
```diff
@@ -4,296 +4,324 @@ sidebar_position: 1
...
-:::danger Not Acceptable
-This design is terrible.
-:::
+## General guidelines and tips
+
+This section provides some general guidelines for the Taichi community and tips that we find practically useful. 
+
+### Be pragmatic & no overkills
+
+Always use straightforward solutions for easy problems so that you have time for really hard ones.
```
[the remark this answers is not in the document]
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:19]: Always use straightforward (sans "the") (sometimes even brute-force) solutions: Complicated code usually suggests a lack of design *or* over-engineering.

File: docs/lang/articles/contribution/contributor_guide.md [hunk 8]
```diff
@@ -4,296 +4,324 @@ sidebar_position: 1
...
+
+> "There are two ways of constructing a software design: One way is to make it so simple that there are obviously no deficiencies, and the other way is to make it so complicated that there are no obvious deficiencies. *The first method is far more difficult*." — [C.A.R. Hoare](https://en.wikipedia.org/wiki/Tony_Hoare)
+
+### Juxtapose pros and cons
```
[the remark this answers is not in the document]
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:23]: Tongue-in-cheek.gif

File: docs/lang/articles/contribution/contributor_guide.md [hunk 9]
```diff
@@ -4,296 +4,324 @@ sidebar_position: 1
...
+
+When it comes to making a design decision, weigh up its pros and cons. A design is *good to go,* so long as its advantages outweigh its disadvantages.
```
[the remark this answers is not in the document]
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:25]: Sounds good.

File: docs/lang/articles/contribution/contributor_guide.md (hunk 4 above)
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:283]: Agreed.

File: docs/lang/articles/contribution/contributor_guide.md (hunk 2 above)
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:227]: @ailzhang Please confirm if we need CPU, LLVM, Metal, Vulkan, OpenGL, Std, Opt, and GUI, **and** if we need "misc" if anything does not fit into the existing categories.

Comment (ailzhang) [reply docs/lang/articles/contribution/contributor_guide.md:227]: Yea let's just keep the most common ones here :D

File: docs/lang/articles/contribution/contributor_guide.md (hunk 3 above)
Comment (ailzhang) [reply docs/lang/articles/contribution/contributor_guide.md:269]: @k-ye IIUC we still distinguish upper case and lower case but there's no enforcement on that :P So let's just keep this line here as suggested action? ;)

Comment (strongoier) [reply docs/lang/articles/contribution/contributor_guide.md:230]: We still distinguish them in the `Highlights` section of the version log. However, it seems that we don't copy them to GitHub for release notes now.

**Review by ailzhang — APPROVED** [review]
Going to merge this PR for now so that we can update the website.  But feel free to keep reviewing this PR and @Vissidarte-Herman can resolve them in followup PRs.
